### ISSUE TITLE
Split ME3Explorer to UI + Library projects

### ISSUE THREAD

**Issue by Mgamerz** [ME3Tweaks/LegendaryExplorer#188]
As ME3Explorer provides most of the modding scene's package handling code (the other project that I know of that provides this to any degree is Mass Effect Modder, however it doesn't provide struct property parsing), I think it would be a good move to split ME3Explorer as it currently stands into two parts, one for the UI that we use for the app and the backend library that can be used by other developers. There is already an existing project for this https://github.com/ME3Tweaks/ME3ExplorerLib but it is based on 4.0, which is very old by our current codebase standards.

I know there is the ME2 custom class builder that needs package handling code that is not provided by the 4.0 based library due to recent discoveries of ME2 properties. ALOT Installer v4 will need to be able to patch a few package files as well to overcome some issues with a few packages.

To implement this kind of change, we would need to identify several goals for each project.

- What should remain in the UI-app (ME3Explorer)?
- What should be placed into the library (ME3ExplorerCore)?

It is my ideas that:

- **UI**: All current UI tools will remain in the UI-app. This includes things like experiments, convenience things, and automated tasks, such as compacting a TFC or AFC file and sequence editing logic, for example. Settings would remain here. ASI Manager would remain here too. Audio and texture handling code (such as WwiseHelper, ISBankEntry) should remain in ME3Explorer UI because these depend on many external things like vgmstream and CompressionCodecs which are not easily portable.
 - **Library**: Package handling, DLC handling, property parsing, and 'back end' logic should be moved here, such as script decompilation (if possible). From my work on M3 and AICore I've also identified the need to minimize dependencies, or merge them into the library and remove them from the hosting app. We should attempt to minimize our dependency on the Windows specific APIs if possible to allow this to work cross platform. Enums like MEGame would also be here.

I am not sure where we would put texture handling.

I believe the library namespaces should be changed to ME3ExplorerCore, while the main UI app remains on ME3Explorer. The target framework for the library would be .netstandard~~2.1~~ 2.0 as it the last one that will support .NET 4.8.

I also think for the library we should attempt to implement https://github.com/Fody/PropertyChanged. This package applies at build time and saves HUGE amounts of boilerplate and provides almost all of the property changed events we heavily depend on, as well as providing very useful accessory methods like `On<propertyname>Changed(){ }` automatically being built and hooked up for us. With this feature we could remove tons of ugly boilerplate, however the tradeoff is we will need to ensure we don't just 'remove unused methods', as they will only be linked to at compile. 

As a library project, we would need to make sure we also document the public facing API.

Thoughts on this?

**Comment by mirh** [comment]
Idk about any technical detail, but I'd just like to remember my suggestion of tools being usable to some extent from the command line.

**Comment by Mgamerz** [comment]
You would have to build a command line wrapper. The library I speak of here would be a .NET dll that is linked.

**Comment by Mgamerz** [comment]
I've begun some discovery work on this project and have so far found the following:

1) We have tons of UI code mixed with the backend. It will be considerable effort to unweave these items. I'm looking at DLC unpacker for instance and there's dependenices on the unpacker UI, as well as lots of things like statuses which should be done through callbacks. I've ported in Unreal, several helpers, and the packages folder. There is a lot of dependency on SharpDX's vector structs, which I am not sure how we will work around as SharpDX is definitely windows only. I am not sure how easy it would be to weed out things that depend on SharpDX and not, as the serialization system @SirCxyrtyx has built includes these items in them and are heavily used by things like Meshplorer and Package Editor.

2) In order to work with .NET Framework 4.8 we can only use .NET Standard 2.0, not 2.1, as 2.1 only works on .NET Core ~~2.2~~ 3.0 (I think?). 2.0 only works with C# 7.3 and not 8.0, so some of  our existing code does not work. We would have to rewrite it to use 2.0, or upgrade the main application to use .NET Core. I am personally not opposed to this, however it introduces other complications, such as significantly increased filesizes of the app unless we want to tie ourselves to a version of .NET core that the user must install in order to use the program. Right now the main features missing are the `using` declarations that don't use a scope, so not too hard to work around.

![image](https://user-images.githubusercontent.com/2738836/91650214-09b3c300-ea3a-11ea-8227-48e902e5cfa6.png)

**Comment by Mgamerz** [comment]
Looking at this, it looks like we can still use C# 8 as long as we avoid using things like nullable (I don't think we use this), async streams (we don't use this). The special switch statements and other stuff will work however as they are just syntax and not runtime features.

**Comment by SirCxyrtyx** [comment]
ME3Explorer uses 4.8, so all of its code is compatible with 4.8.

Is making it work cross-platform necessary? Mass Effect games are windows only. (I'm aware that some people manage to run them on linux somehow, but would whatever emulation or vm software they're using to do that not work for this?)

**Comment by Mgamerz** [comment]
ALOT Installer v4 will work cross platform as the main developers and
several of the more advanced users use linux. I think it would be useful to
try and make it cross plat if possible. I've built in some shims for
sharpdx from the source code that make most of it already work. The only
thing it needs now is some deconstructors so the serialization can work.
There's a few other thing but most of the core library is actually moved in
the library split branch. (Currently there's a copy of sharpdx mathematics
for testing purposes but will be removed).

**Comment by mirh** [comment]
Wine is neither virtualization nor emulation. 

Said this, it doesn't really make any sense to focus on a native version when the game would still eventually require it. 
Having at least the "apply mods" tools to run on XP would probably help more people.

**Comment by Mgamerz** [comment]
We are talking about ME3Explorer, not mod manager. ME3Explorer is the developer toolset and does not install mods nor is it designed to be used by the same userbase.

A library version of the toolset backend would be used by other tools (as well as ME3Explorer front end), which so far would include alot installer v4 (linux/windows/mac, since .net core supports all three and I have a console version ready. Not sure I will take the time to test mac though), and the custom class builder which is windows wpf, they also need an updated version of the ME3Explorer lib. The current version is 4.0 and is really outdated and doesn't even properly work on me2.

It also makes it much easier to build other distributable tools that a developer may want, for example I made a version a version a while ago to fix the issues on the fly for PEOM when it had the null name issue, and for patching the mars to citadel transition. The library version would make it much easier to keep up to date with toolset advancements such as relinking and improved property parsing.

We have linux and mac versions of the compression libraries available, as we use the mem implementations of the compression  already, and aquadran has provided me the linux libraries already.

To be clear this is only the backend that I am looking to see if it could be made cross plat capable. The front end would be 100% windows only on .Net Framework 4.8. Currently it is about 80% of the way there - though I haven't linked the main app to it yet. This also conveniently moves all of the helpers project into the core library, with helpers essentially being image and texture related. I have removed a bunch of ancient package code from the PCCObject days when we used managed decompression instead of our current native one.
